File: src/studio/api/pipelines.py
"""
Pipeline API Routes

FastAPI routes for pipeline orchestration management.
"""

from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel, Field

from studio.middleware.rbac import get_current_user_from_request
from studio.services.pipeline_service import ORCHESTRATION_PATTERNS, PipelineService
from studio.services.rbac_service import RBACService

router = APIRouter(prefix="/pipelines", tags=["Pipelines"])


# ===================
# Request/Response Models
# ===================


class PipelineCreate(BaseModel):
    """Request model for creating a pipeline."""

    organization_id: str
    workspace_id: str
    name: str
    pattern: str = Field(
        ...,
        description="Orchestration pattern: sequential, parallel, router, supervisor, ensemble",
    )
    description: str = ""
    status: str = "draft"


class PipelineUpdate(BaseModel):
    """Request model for updating a pipeline."""

    name: str | None = None
    description: str | None = None
    pattern: str | None = None
    status: str | None = None


class NodeCreate(BaseModel):
    """Request model for creating a pipeline node."""

    id: str | None = None
    node_type: str = Field(
        ..., description="Node type: agent, input, output, condition, merge"
    )
    agent_id: str | None = ""
    label: str
    position_x: float = 0.0
    position_y: float = 0.0
    config: dict | None = None


class NodeUpdate(BaseModel):
    """Request model for updating a pipeline node."""

    node_type: str | None = None
    agent_id: str | None = None
    label: str | None = None
    position_x: float | None = None
    position_y: float | None = None
    config: dict | None = None


class ConnectionCreate(BaseModel):
    """Request model for creating a connection."""

    id: str | None = None
    source_node_id: str
    target_node_id: str
    source_handle: str = "output"
    target_handle: str = "input"
    condition: dict | None = None


class GraphSave(BaseModel):
    """Request model for saving complete graph."""

    nodes: list[NodeCreate]
    connections: list[ConnectionCreate]


# ===================
# Dependencies
# ===================


def get_pipeline_service() -> PipelineService:
    """Get pipeline service instance."""
    return PipelineService()


def get_rbac_service() -> RBACService:
    """Get RBAC service instance."""
    return RBACService()


async def check_permission(
    user: dict,
    permission: str,
    rbac_service: RBACService,
):
    """Check if user has required permission."""
    has_perm = await rbac_service.check_permission(
        user_id=user["id"],
        permission=permission,
    )
    if not has_perm:
        raise HTTPException(status_code=403, detail=f"Permission denied: {permission}")


# ===================
# Pipeline CRUD Routes
# ===================


@router.post("", response_model=dict)
async def create_pipeline(
    data: PipelineCreate,
    user: dict = Depends(get_current_user_from_request),
    service: PipelineService = Depends(get_pipeline_service),
    rbac_service: RBACService = Depends(get_rbac_service),
):
    """
    Create a new pipeline.

    Requires: agents:create permission
    """
    await check_permission(user, "agents:create", rbac_service)

    try:
        pipeline = await service.create(
            organization_id=data.organization_id,
            workspace_id=data.workspace_id,
            name=data.name,
            pattern=data.pattern,
            created_by=user["id"],
            description=data.description,
            status=data.status,
        )
        return {"data": pipeline}
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.get("", response_model=dict)
async def list_pipelines(
    organization_id: str = Query(..., description="Organization ID"),
    workspace_id: str | None = Query(None, description="Workspace ID filter"),
    status: str | None = Query(None, description="Status filter"),
    limit: int = Query(50, ge=1, le=100),
    offset: int = Query(0, ge=0),
    user: dict = Depends(get_current_user_from_request),
    service: PipelineService = Depends(get_pipeline_service),
    rbac_service: RBACService = Depends(get_rbac_service),
):
    """
    List pipelines with optional filters.

    Requires: agents:read permission
    """
    await check_permission(user, "agents:read", rbac_service)

    filters = {}
    if status:
        filters["status"] = status

    result = await service.list(
        organization_id=organization_id,
        workspace_id=workspace_id,
        filters=filters if filters else None,
        limit=limit,
        offset=offset,
    )
    return {"data": result["records"], "total": result["total"]}


@router.get("/patterns", response_model=dict)
async def get_orchestration_patterns():
    """
    Get available orchestration patterns.

    No authentication required.
    """
    return {"data": ORCHESTRATION_PATTERNS}


@router.get("/{pipeline_id}", response_model=dict)
async def get_pipeline(
    pipeline_id: str,
    user: dict = Depends(get_current_user_from_request),
    service: PipelineService = Depends(get_pipeline_service),
    rbac_service: RBACService = Depends(get_rbac_service),
):
    """
    Get a pipeline with its graph (nodes and connections).

    Requires: agents:read permission
    """
    await check_permission(user, "agents:read", rbac_service)

    pipeline = await service.get_with_graph(pipeline_id)
    if not pipeline:
        raise HTTPException(status_code=404, detail="Pipeline not found")

    return {"data": pipeline}


@router.put("/{pipeline_id}", response_model=dict)
async def update_pipeline(
    pipeline_id: str,
    data: PipelineUpdate,
    user: dict = Depends(get_current_user_from_request),
    service: PipelineService = Depends(get_pipeline_service),
    rbac_service: RBACService = Depends(get_rbac_service),
):
    """
    Update a pipeline.

    Requires: agents:update permission
    """
    await check_permission(user, "agents:update", rbac_service)

    # Build update dict from non-None fields
    update_data = {k: v for k, v in data.model_dump().items() if v is not None}
    if not update_data:
        raise HTTPException(status_code=400, detail="No fields to update")

    try:
        pipeline = await service.update(pipeline_id, update_data)
        if not pipeline:
            raise HTTPException(status_code=404, detail="Pipeline not found")
        return {"data": pipeline}
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.delete("/{pipeline_id}", response_model=dict)
async def delete_pipeline(
    pipeline_id: str,
    user: dict = Depends(get_current_user_from_request),
    service: PipelineService = Depends(get_pipeline_service),
    rbac_service: RBACService = Depends(get_rbac_service),
):
    """
    Archive a pipeline (soft delete).

    Requires: agents:delete permission
    """
    await check_permission(user, "agents:delete", rbac_service)

    pipeline = await service.get(pipeline_id)
    if not pipeline:
        raise HTTPException(status_code=404, detail="Pipeline not found")

    await service.delete(pipeline_id)
    return {"message": "Pipeline archived"}


# ===================
# Graph Operations
# ===================


@router.put("/{pipeline_id}/graph", response_model=dict)
async def save_graph(
    pipeline_id: str,
    data: GraphSave,
    user: dict = Depends(get_current_user_from_request),
    service: PipelineService = Depends(get_pipeline_service),
    rbac_service: RBACService = Depends(get_rbac_service),
):
    """
    Save complete pipeline graph (replaces existing nodes and connections).

    Requires: agents:update permission
    """
    await check_permission(user, "agents:update", rbac_service)

    pipeline = await service.get(pipeline_id)
    if not pipeline:
        raise HTTPException(status_code=404, detail="Pipeline not found")

    # Convert Pydantic models to dicts
    nodes = [node.model_dump() for node in data.nodes]
    connections = [conn.model_dump() for conn in data.connections]

    result = await service.save_graph(pipeline_id, nodes, connections)
    return {"data": result}


@router.post("/{pipeline_id}/validate", response_model=dict)
async def validate_pipeline(
    pipeline_id: str,
    user: dict = Depends(get_current_user_from_request),
    service: PipelineService = Depends(get_pipeline_service),
    rbac_service: RBACService = Depends(get_rbac_service),
):
    """
    Validate pipeline graph structure.

    Requires: agents:read permission
    """
    await check_permission(user, "agents:read", rbac_service)

    pipeline = await service.get(pipeline_id)
    if not pipeline:
        raise HTTPException(status_code=404, detail="Pipeline not found")

    result = await service.validate(pipeline_id)
    return {"data": result}


# ===================
# Node Operations
# ===================


@router.get("/{pipeline_id}/nodes", response_model=dict)
async def list_nodes(
    pipeline_id: str,
    user: dict = Depends(get_current_user_from_request),
    service: PipelineService = Depends(get_pipeline_service),
    rbac_service: RBACService = Depends(get_rbac_service),
):
    """
    List all nodes in a pipeline.

    Requires: agents:read permission
    """
    await check_permission(user, "agents:read", rbac_service)

    pipeline = await service.get(pipeline_id)
    if not pipeline:
        raise HTTPException(status_code=404, detail="Pipeline not found")

    nodes = await service.list_nodes(pipeline_id)
    return {"data": nodes}


@router.post("/{pipeline_id}/nodes", response_model=dict)
async def add_node(
    pipeline_id: str,
    data: NodeCreate,
    user: dict = Depends(get_current_user_from_request),
    service: PipelineService = Depends(get_pipeline_service),
    rbac_service: RBACService = Depends(get_rbac_service),
):
    """
    Add a node to a pipeline.

    Requires: agents:update permission
    """
    await check_permission(user, "agents:update", rbac_service)

    pipeline = await service.get(pipeline_id)
    if not pipeline:
        raise HTTPException(status_code=404, detail="Pipeline not found")

    node = await service.add_node(pipeline_id, data.model_dump())
    return {"data": node}


@router.put("/{pipeline_id}/nodes/{node_id}", response_model=dict)
async def update_node(
    pipeline_id: str,
    node_id: str,
    data: NodeUpdate,
    user: dict = Depends(get_current_user_from_request),
    service: PipelineService = Depends(get_pipeline_service),
    rbac_service: RBACService = Depends(get_rbac_service),
):
    """
    Update a pipeline node.

    Requires: agents:update permission
    """
    await check_permission(user, "agents:update", rbac_service)

    # Verify pipeline exists
    pipeline = await service.get(pipeline_id)
    if not pipeline:
        raise HTTPException(status_code=404, detail="Pipeline not found")

    # Build update dict from non-None fields
    update_data = {k: v for k, v in data.model_dump().items() if v is not None}
    if not update_data:
        raise HTTPException(status_code=400, detail="No fields to update")

    node = await service.update_node(node_id, update_data)
    if not node:
        raise HTTPException(status_code=404, detail="Node not found")

    return {"data": node}


@router.delete("/{pipeline_id}/nodes/{node_id}", response_model=dict)
async def remove_node(
    pipeline_id: str,
    node_id: str,
    user: dict = Depends(get_current_user_from_request),
    service: PipelineService = Depends(get_pipeline_service),
    rbac_service: RBACService = Depends(get_rbac_service),
):
    """
    Remove a node from a pipeline.

    Requires: agents:update permission
    """
    await check_permission(user, "agents:update", rbac_service)

    # Verify pipeline exists
    pipeline = await service.get(pipeline_id)
    if not pipeline:
        raise HTTPException(status_code=404, detail="Pipeline not found")

    # Verify node exists
    node = await service.get_node(node_id)
    if not node:
        raise HTTPException(status_code=404, detail="Node not found")

    await service.remove_node(node_id)
    return {"message": "Node removed"}


# ===================
# Connection Operations
# ===================


@router.get("/{pipeline_id}/connections", response_model=dict)
async def list_connections(
    pipeline_id: str,
    user: dict = Depends(get_current_user_from_request),
    service: PipelineService = Depends(get_pipeline_service),
    rbac_service: RBACService = Depends(get_rbac_service),
):
    """
    List all connections in a pipeline.

    Requires: agents:read permission
    """
    await check_permission(user, "agents:read", rbac_service)

    pipeline = await service.get(pipeline_id)
    if not pipeline:
        raise HTTPException(status_code=404, detail="Pipeline not found")

    connections = await service.list_connections(pipeline_id)
    return {"data": connections}


@router.post("/{pipeline_id}/connections", response_model=dict)
async def add_connection(
    pipeline_id: str,
    data: ConnectionCreate,
    user: dict = Depends(get_current_user_from_request),
    service: PipelineService = Depends(get_pipeline_service),
    rbac_service: RBACService = Depends(get_rbac_service),
):
    """
    Add a connection between nodes.

    Requires: agents:update permission
    """
    await check_permission(user, "agents:update", rbac_service)

    pipeline = await service.get(pipeline_id)
    if not pipeline:
        raise HTTPException(status_code=404, detail="Pipeline not found")

    connection = await service.add_connection(pipeline_id, data.model_dump())
    return {"data": connection}


@router.delete("/{pipeline_id}/connections/{connection_id}", response_model=dict)
async def remove_connection(
    pipeline_id: str,
    connection_id: str,
    user: dict = Depends(get_current_user_from_request),
    service: PipelineService = Depends(get_pipeline_service),
    rbac_service: RBACService = Depends(get_rbac_service),
):
    """
    Remove a connection.

    Requires: agents:update permission
    """
    await check_permission(user, "agents:update", rbac_service)

    # Verify pipeline exists
    pipeline = await service.get(pipeline_id)
    if not pipeline:
        raise HTTPException(status_code=404, detail="Pipeline not found")

    # Verify connection exists
    connection = await service.get_connection(connection_id)
    if not connection:
        raise HTTPException(status_code=404, detail="Connection not found")

    await service.remove_connection(connection_id)
    return {"message": "Connection removed"}

File: apps/web/src/features/pipelines/api/pipelines.ts
import apiClient from "@/api";
import type { Pipeline, PipelinePattern } from "../types";

export interface PipelineFilters {
  organization_id?: string;
  workspace_id?: string;
  status?: "draft" | "active" | "archived";
  pattern?: PipelinePattern;
  search?: string;
  limit?: number;
  offset?: number;
}

export interface CreatePipelineInput {
  organization_id: string;
  workspace_id: string;
  name: string;
  description?: string;
  pattern: PipelinePattern;
}

export interface UpdatePipelineInput {
  name?: string;
  description?: string;
  pattern?: PipelinePattern;
  status?: "draft" | "active" | "archived";
}

/**
 * Backend node format for saving graph
 */
export interface BackendNodeCreate {
  id?: string;
  node_type: string;
  agent_id?: string;
  label: string;
  position_x: number;
  position_y: number;
  config?: Record<string, unknown>;
}

/**
 * Backend connection format for saving graph
 */
export interface BackendConnectionCreate {
  id?: string;
  source_node_id: string;
  target_node_id: string;
  source_handle?: string;
  target_handle?: string;
  condition?: Record<string, unknown>;
}

/**
 * Input for saving graph (nodes and connections)
 */
export interface SaveGraphInput {
  nodes: BackendNodeCreate[];
  connections: BackendConnectionCreate[];
}

/**
 * Backend node type (from API)
 */
interface BackendNode {
  id: string;
  node_type: string;
  agent_id?: string;
  label: string;
  position_x: number;
  position_y: number;
  config?: Record<string, unknown>;
}

/**
 * Backend connection type (from API)
 */
interface BackendConnection {
  id: string;
  source_node_id: string;
  target_node_id: string;
  source_handle?: string;
  target_handle?: string;
  condition?: Record<string, unknown>;
}

/**
 * Transform backend node to frontend PipelineNode type
 */
function transformNode(backendNode: BackendNode): Pipeline["nodes"][0] {
  return {
    id: backendNode.id,
    type: backendNode.node_type as Pipeline["nodes"][0]["type"],
    position: {
      x: backendNode.position_x,
      y: backendNode.position_y,
    },
    data: {
      label: backendNode.label,
      agentId: backendNode.agent_id,
      config: backendNode.config,
    },
  };
}

/**
 * Transform backend connection to frontend PipelineEdge type
 */
function transformConnection(
  backendConnection: BackendConnection
): Pipeline["edges"][0] {
  return {
    id: backendConnection.id,
    source: backendConnection.source_node_id,
    target: backendConnection.target_node_id,
    sourceHandle: backendConnection.source_handle,
    targetHandle: backendConnection.target_handle,
    condition: backendConnection.condition?.expression as string | undefined,
  };
}

/**
 * Transform backend pipeline to frontend Pipeline type
 */
function transformPipeline(backendPipeline: Record<string, unknown>): Pipeline {
  // Backend uses 'connections' but frontend expects 'edges'
  const backendNodes = (backendPipeline.nodes as BackendNode[]) || [];
  const backendConnections =
    (backendPipeline.connections as BackendConnection[]) ||
    (backendPipeline.edges as BackendConnection[]) ||
    [];

  // Transform nodes from backend format to frontend format
  const nodes = backendNodes.map(transformNode);

  // Transform connections from backend format to frontend edges format
  const edges = backendConnections.map(transformConnection);

  return {
    id: backendPipeline.id as string,
    name: (backendPipeline.name as string) || "Untitled Pipeline",
    description: (backendPipeline.description as string) || "",
    pattern: (backendPipeline.pattern as PipelinePattern) || "sequential",
    nodes,
    edges,
    status: (backendPipeline.status as Pipeline["status"]) || "draft",
    organization_id: (backendPipeline.organization_id as string) || "",
    created_by: (backendPipeline.created_by as string) || "",
    created_at: backendPipeline.created_at as string,
    updated_at: backendPipeline.updated_at as string,
  };
}

export interface PipelinesListResponse {
  items: Pipeline[];
  total: number;
}

export const pipelinesApi = {
  /**
   * Get all pipelines with optional filters
   */
  getAll: async (filters?: PipelineFilters): Promise<PipelinesListResponse> => {
    const response = await apiClient.get<
      | {
          data: Record<string, unknown>[];
          records: Record<string, unknown>[];
          total: number;
        }
      | Record<string, unknown>[]
    >("/api/v1/pipelines", { params: filters });

    // Handle both array and paginated response formats
    if (Array.isArray(response.data)) {
      const pipelines = response.data.map(transformPipeline);
      return { items: pipelines, total: pipelines.length };
    }

    // Backend returns 'data' array, but handle 'records' for compatibility
    const pipelinesArray = response.data.data || response.data.records || [];
    const pipelines = pipelinesArray.map(transformPipeline);
    return { items: pipelines, total: response.data.total || pipelines.length };
  },

  /**
   * Get a single pipeline by ID
   */
  getById: async (id: string): Promise<Pipeline> => {
    const response = await apiClient.get<
      { data: Record<string, unknown> } | Record<string, unknown>
    >(`/api/v1/pipelines/${id}`);
    // Backend wraps response in 'data' field
    const pipelineData =
      (response.data as { data: Record<string, unknown> }).data ||
      response.data;
    return transformPipeline(pipelineData as Record<string, unknown>);
  },

  /**
   * Create a new pipeline
   */
  create: async (data: CreatePipelineInput): Promise<Pipeline> => {
    const response = await apiClient.post<Record<string, unknown>>(
      "/api/v1/pipelines",
      data
    );
    return transformPipeline(response.data);
  },

  /**
   * Update an existing pipeline
   */
  update: async (id: string, data: UpdatePipelineInput): Promise<Pipeline> => {
    const response = await apiClient.put<Record<string, unknown>>(
      `/api/v1/pipelines/${id}`,
      data
    );
    return transformPipeline(response.data);
  },

  /**
   * Delete a pipeline
   */
  delete: async (id: string): Promise<void> => {
    await apiClient.delete(`/api/v1/pipelines/${id}`);
  },

  /**
   * Duplicate a pipeline
   */
  duplicate: async (id: string): Promise<Pipeline> => {
    const response = await apiClient.post<Record<string, unknown>>(
      `/api/v1/pipelines/${id}/duplicate`
    );
    return transformPipeline(response.data);
  },

  /**
   * Save pipeline graph (nodes and connections)
   * This is separate from update() because the backend uses different endpoints
   */
  saveGraph: async (
    id: string,
    nodes: Pipeline["nodes"],
    edges: Pipeline["edges"]
  ): Promise<{ nodes_saved: number; connections_saved: number }> => {
    // Transform frontend format to backend format
    const backendNodes: BackendNodeCreate[] = nodes.map((node) => ({
      id: node.id,
      node_type: node.type,
      agent_id: node.data.agentId || "",
      label: node.data.label,
      position_x: node.position.x,
      position_y: node.position.y,
      config: node.data.config,
    }));

    const backendConnections: BackendConnectionCreate[] = edges.map((edge) => ({
      id: edge.id,
      source_node_id: edge.source,
      target_node_id: edge.target,
      source_handle: edge.sourceHandle || "output",
      target_handle: edge.targetHandle || "input",
      condition: edge.condition ? { expression: edge.condition } : undefined,
    }));

    const response = await apiClient.put<{ data: Record<string, unknown> }>(
      `/api/v1/pipelines/${id}/graph`,
      {
        nodes: backendNodes,
        connections: backendConnections,
      }
    );

    // Backend returns the full pipeline - extract counts from the returned data
    const pipelineData = response.data.data || response.data;
    const returnedNodes = (pipelineData.nodes as BackendNode[]) || [];
    const returnedConnections =
      (pipelineData.connections as BackendConnection[]) || [];

    return {
      nodes_saved: returnedNodes.length,
      connections_saved: returnedConnections.length,
    };
  },

  /**
   * Validate a pipeline's graph structure
   */
  validate: async (
    id: string
  ): Promise<{ valid: boolean; errors?: string[] }> => {
    const response = await apiClient.post<{
      data: { valid: boolean; errors?: string[] };
    }>(`/api/v1/pipelines/${id}/validate`);
    return response.data.data;
  },
};

export default pipelinesApi;
